### ISSUE TITLE
[BUG] Non-class data member before `this`

### ISSUE THREAD

**Issue by JohelEGP** [hsutter/cppfront#332]
`cpp2::store_as_base<T>` inherits from `T`, which only works for `class` types. I suggest its API follows <https://www.boost.org/doc/libs/release/libs/utility/doc/html/utility/utilities/base_from_member.html>.

**Minimal reproducer**:
```Cpp2
x: type = { }
y: type = {
  a: int;
  this: x;
}
```

Commands:
```bash
cppfront x.cpp2
clang++17 -std=c++2b -I $CPPFRONT_INCLUDE_DIR x.cpp
```

Expected result: A well-formed program.

Actual result and error:

<details><summary>Generated C++1</summary>

```C++


//=== Cpp2 type declarations ====================================================


#include "cpp2util.h"

#line 1 "x.cpp2"
class x;
class y;

//=== Cpp2 type definitions and function declarations ===========================

#line 1 "x.cpp2"
class x {};
class y: private cpp2::store_as_base<int>, public x {

#line 5 "x.cpp2"
};

//=== Cpp2 function definitions =================================================



```

</details>

```output
$CPPFRONT_INCLUDE_DIR/cpp2util.h:623:32: error: base specifier must name a class
struct store_as_base : private T
                       ~~~~~~~~^
x.cpp2:2:18: note: in instantiation of template class 'cpp2::store_as_base<int>' requested here
class y: private cpp2::store_as_base<int>, public x {
                 ^
1 error generated.
```

**Comment by JohelEGP** [comment]
0982b8ec41f7ed880d748269c7732d56228a4a19's commit message mentions
> In this initial checkin, if there are any non-base subobjects (ordinary data members) that are declared before base subobjects (base classes), as an implementation detail those non-bases are emitted as private base classes via a helper wrapper that prevents their interface (functions) from leaking into the type.

That seems like the justification for private inheritance. But having the `T` be a data member works just as well.

**Comment by hsutter** [comment]
Thanks! Yes, I think this is now fixed with the same commit 4f689620084caab158bcea427d9f18ffbd135b75 that fixed #333, and indeed the thinko was "d'oh why did I use a private base? just hold it as a member"... please reopen if that hasn't fixed it.
